org.mixedrealitytoolkit.spatialmanipulation/Solvers/TapToPlace.cs

```diff
@@ -577,6 +577,10 @@ private void RegisterPlacementAction()
                         }
                     }
 #pragma warning restore CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
+                    else //if controllerInteractor.xrController is null then it is an interactor from a controllerless controller
```

Comment:
The registration and unregistration code needs to be cleaned up.  We're missing some case here.  In particular if the interactor uses a "action" instead of an "action reference".    Also in all cases we're missing the fallback case where we fallback to the "selectEnter" event.  Removing the need for the assert.

Here's the code:

```cs
        /// <summary>
        /// Registers the input action which performs placement.
        /// </summary>
        private void RegisterPlacementAction()
        {
            // Refresh the registeration if they already exist
            UnregisterPlacementAction();

            if (interactionManager == null)
            {
                interactionManager = ComponentCache<XRInteractionManager>.FindFirstActiveInstance();
                if (interactionManager == null)
                {
                    Debug.LogError("No interaction manager found in scene. Please add an interaction manager to the scene.");
                }
            }

            if (interactorsCache == null)
            {
                interactorsCache = new List<IXRInteractor>();
            }

            // Try registering for the controller's "action" so object selection isn't required for placement.
            // If no controller, then fallback to using object selections for placement.
            interactionManager.GetRegisteredInteractors(interactorsCache);
            foreach (IXRInteractor interactor in interactorsCache)
            {
#pragma warning disable CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
                if (interactor is XRBaseControllerInteractor controllerInteractor &&
                    controllerInteractor.xrController is ActionBasedController actionController)
                {
                    actionController.selectAction.action.performed += StopPlacementViaPerformedAction;
                }
#pragma warning restore CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
                else if (interactor is XRBaseInputInteractor inputInteractor &&
                    TryGetActionValueOrActionReferenceValue(inputInteractor.selectInput, out InputAction selectionAction))
                {
                    selectionAction.performed += StopPlacementViaPerformedAction;

                }                  
                else if (interactor is IXRSelectInteractor selectInteractor)
                {
                    selectInteractor.selectEntered.AddListener(StopPlacementViaSelect);
                }
            }
        }

        /// <summary>
        /// Unregisters the input action which performs placement.
        /// </summary>
        private void UnregisterPlacementAction()
        {
            if (interactorsCache != null)
            {
                foreach (IXRInteractor interactor in interactorsCache)
                {
#pragma warning disable CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
                    if (interactor is XRBaseControllerInteractor controllerInteractor &&
                        controllerInteractor.xrController is ActionBasedController actionController)
                    {
                        actionController.selectAction.action.performed -= StopPlacementViaPerformedAction;
                    }
#pragma warning restore CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
                    else if (interactor is XRBaseInputInteractor inputInteractor &&
                        TryGetActionValueOrActionReferenceValue(inputInteractor.selectInput, out InputAction selectionAction))
                    {
                        selectionAction.performed -= StopPlacementViaPerformedAction;

                    }
                    else if (interactor is IXRSelectInteractor selectInteractor)
                    {
                        selectInteractor.selectEntered.RemoveListener(StopPlacementViaSelect);
                    }
                }
                interactorsCache.Clear();
            }
        }

        /// <summary>
        /// Try to obtain the action value or action referecne value from the given button reader.
        /// </summary>
        private bool TryGetActionValueOrActionReferenceValue(XRInputButtonReader reader, out InputAction action)
        {
            if (reader == null)
            {
                action = null;
                return false;
            }

            if (reader.inputActionValue != null)
            {
                action = reader.inputActionValue;
                return true;
            }

            if (reader.inputActionReferenceValue)
            {
                action = reader.inputActionReferenceValue.action;
                return true;
            }

            action = null;
            return false;
        }
```

Reply:
Thank you for your detailed feedback and providing the code.  Unfortunately, those are breaking changes.  I tested them and they break these tests:
<img width="293" alt="image" src="https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/4ad7f8f7-52c5-4f8b-a33d-8c4c7622a549">
Maybe you have other changes in your branch that I would have to add too?
In the meantime, I won't add them to this PR.  Resolving, please feel free to re-open if needed.

Reply:
hmmm...ok let's figure this out.  This code needs updating.  it's missing some cases that the old non-XRI3 covered.

Did you investigate the unit test failures?

Reply:
Not yet, got distracted on other tasks.  I'll look into those in a little while.

Reply:
Ok I had missed some stuff on the `TryGetActionValueOrActionReferenceValue` function... try updating my original code with this version of that function.

```cs
        /// <summary>
        /// Try to obtain the action value or action referecne value from the given button reader.
        /// </summary>
        private bool TryGetActionValueOrActionReferenceValue(XRInputButtonReader reader, out InputAction action)
        {
            if (reader == null)
            {
                action = null;
                return false;
            }

            if (reader.inputSourceMode == XRInputButtonReader.InputSourceMode.InputAction &&
                reader.inputActionPerformed != null)
            {
                action = reader.inputActionPerformed;
                return true;
            }

            if (reader.inputSourceMode == XRInputButtonReader.InputSourceMode.InputActionReference &&
                reader.inputActionReferencePerformed)
            {
                action = reader.inputActionReferencePerformed.action;
                return true;
            }

            action = null;
            return false;
        }
```

Reply:
Actually  I had other bugs....here's the entire change

```cs
/// <summary>
/// Registers the input action which performs placement.
/// </summary>
private void RegisterPlacementAction()
{
    // Refresh the registeration if they already exist
    UnregisterPlacementAction();

    if (interactionManager == null)
    {
        interactionManager = ComponentCache<XRInteractionManager>.FindFirstActiveInstance();
        if (interactionManager == null)
        {
            Debug.LogError("No interaction manager found in scene. Please add an interaction manager to the scene.");
        }
    }

    if (interactorsCache == null)
    {
        interactorsCache = new List<IXRInteractor>();
    }

    // Try registering for the controller's "action" so object selection isn't required for placement.
    // If no controller, then fallback to using object selections for placement.
    interactionManager.GetRegisteredInteractors(interactorsCache);
    foreach (IXRInteractor interactor in interactorsCache)
    {
#pragma warning disable CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
        if (interactor is XRBaseInputInteractor controllerInteractor &&
            controllerInteractor.forceDeprecatedInput &&
            controllerInteractor.xrController is ActionBasedController actionController)
        {
            actionController.selectAction.action.performed += StopPlacementViaPerformedAction;
        }
#pragma warning restore CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
        else if (interactor is XRBaseInputInteractor inputInteractor &&
            TryGetActionValueOrActionReferenceValue(inputInteractor.selectInput, out InputAction selectionAction))
        {
            selectionAction.performed += StopPlacementViaPerformedAction;
        }
        else if (interactor is IXRSelectInteractor selectInteractor)
        {
            selectInteractor.selectEntered.AddListener(StopPlacementViaSelect);
        }
    }
}

/// <summary>
/// Unregisters the input action which performs placement.
/// </summary>
private void UnregisterPlacementAction()
{
    if (interactorsCache != null)
    {
        foreach (IXRInteractor interactor in interactorsCache)
        {
#pragma warning disable CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
            if (interactor is XRBaseInputInteractor controllerInteractor &&
                controllerInteractor.forceDeprecatedInput &&
                controllerInteractor.xrController is ActionBasedController actionController)
            {
                actionController.selectAction.action.performed -= StopPlacementViaPerformedAction;
            }
#pragma warning restore CS0618 // ActionBasedController and XRBaseInputInteractor.forceDeprecatedInput are obsolete
            else if (interactor is XRBaseInputInteractor inputInteractor &&
                TryGetActionValueOrActionReferenceValue(inputInteractor.selectInput, out InputAction selectionAction))
            {
                selectionAction.performed -= StopPlacementViaPerformedAction;
            }
            else if (interactor is IXRSelectInteractor selectInteractor)
            {
                selectInteractor.selectEntered.RemoveListener(StopPlacementViaSelect);
            }
        }
        interactorsCache.Clear();
    }
}

/// <summary>
/// Try to obtain the action value or action referecne value from the given button reader.
/// </summary>
private bool TryGetActionValueOrActionReferenceValue(XRInputButtonReader reader, out InputAction action)
{
    if (reader == null)
    {
        action = null;
        return false;
    }

    if (reader.inputSourceMode == XRInputButtonReader.InputSourceMode.InputAction &&
        reader.inputActionPerformed != null)
    {
        action = reader.inputActionPerformed;
        return true;
    }

    if (reader.inputSourceMode == XRInputButtonReader.InputSourceMode.InputActionReference &&
        reader.inputActionReferencePerformed)
    {
        action = reader.inputActionReferencePerformed.action;
        return true;
    }

    action = null;
    return false;
}
```

Reply:
I added your changes in dc93bd6bd33bd99cff96bb2097542374160fc334, thank you for providing such detailed contribution.  Resolving.